src/crimson/osd/pg_map.h

```diff
@@ -10,12 +10,75 @@
 
 #include "include/types.h"
 #include "crimson/common/type_helpers.h"
+#include "crimson/common/smp_helpers.h"
 #include "crimson/osd/osd_operation.h"
 #include "osd/osd_types.h"
 
 namespace crimson::osd {
 class PG;
 
+/**
+ * PGShardMapping
+ *
+ * Maps pgs to shards.
+ */
+class PGShardMapping {
+public:
+  /// Returns mapping if present, NULL_CORE otherwise
+  core_id_t get_pg_mapping(spg_t pgid) {
+    auto iter = pg_to_core.find(pgid);
+    ceph_assert(iter == pg_to_core.end() || iter->second != NULL_CORE);
+    return iter == pg_to_core.end() ? NULL_CORE : iter->second;
+  }
+
+  /// Returns mapping for pgid, creates new one if it doesn't already exist
+  core_id_t maybe_create_pg(spg_t pgid) {
+    auto [insert_iter, inserted] = pg_to_core.emplace(pgid, NULL_CORE);
+    if (!inserted) {
+      ceph_assert(insert_iter->second != NULL_CORE);
+      return insert_iter->second;
+    } else {
+      auto iter = core_to_num_pgs.begin();
+      auto min_iter = iter++;
+      for (; iter != core_to_num_pgs.end(); ++iter) {
+	if (min_iter->second > iter->second) min_iter = iter;
+      }
+      insert_iter->second = min_iter->first;
+      min_iter->second++;
+      return insert_iter->second;
+    }
+  }
+
+  /// Remove pgid
+  void remove_pg(spg_t pgid) {
+    auto iter = pg_to_core.find(pgid);
+    ceph_assert(iter != pg_to_core.end());
+    ceph_assert(iter->second != NULL_CORE);
+    auto count_iter = core_to_num_pgs.find(iter->second);
+    ceph_assert(count_iter != core_to_num_pgs.end());
+    ceph_assert(count_iter->second > 0);
+    --(count_iter->second);
+    pg_to_core.erase(iter);
+  }
+
+  /// Map to cores in [min_core_mapping, core_mapping_limit)
+  PGShardMapping(core_id_t min_core_mapping, core_id_t core_mapping_limit) {
+    ceph_assert(min_core_mapping < core_mapping_limit);
+    for (auto i = min_core_mapping; i != core_mapping_limit; ++i) {
+      core_to_num_pgs.emplace(i, 0);
```

Comment:
Hmm, I guess single core only (`core_mapping_limit == min_core_mapping + 1`) is allowed here.

Reply:
Swapped for std::min_element

Reply:
We do require at least a single core.